Grant the Okta service the right to write Users

tcsc opened the pull request:
In preparation for the Okta service syncing user accounts, this patch grants the Okta service the right to create/update/delete users, in addition to the existing read permission.

tcsc committed:
Grant the Okta service the right to write Users

In preparation for the Okta service syncing user accounts, this patch grants the Okta
service the right to create/update/delete users, in addition to the existing read
permission.

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

greedy52 commented on lib/authz/permissions.go:
```diff
@@ -916,7 +916,7 @@ func definitionForBuiltinRole(clusterName string, recConfig types.SessionRecordi
 						types.NewRule(types.KindAppServer, services.RW()),
 						types.NewRule(types.KindClusterNetworkingConfig, services.RO()),
-						types.NewRule(types.KindUser, services.RO()),
+						types.NewRule(types.KindUser, services.RW()),
```
RW on KindUser is really powerful. @jentfoo

jentfoo reviewed (changes requested):
Thank you for highlighting @greedy52

I agree this RFD for this change should be reviewed first.  @tcsc please link the RFD when you get a chance and I can take a look.

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

tcsc commented:
Perhaps there is a way we can constrain the write verbs to users that have a `source: okta` label or somesuch?

tcsc commented:
After some experimentation, I'm not sure we are currently able to express such an idea in a Role allow rule. We _can_ certainly check things on the _active_ user in a `where` clause, but I haven't been able to treat a `User` as a generic resource in a rule.

While I can define a rule like

```yaml
allow:
  rules:
  - resources: [user]
    verbs: [create, update, delete]
    where: equals(resource.metadata.labels["teleport.dev/origin"], "okta")
```

which will parse and execute, but never match because `ServerWithRoles` does not specify a `resource` when performing the access check. 

Without extending the user operation access check to include such a concept (which, while I like the idea in principle, needs a _lot_ more thought put into it than this), constraining the Okta role using our normal rule mechanisms seems like a dead end.

Suggestions gratefully accepted.

greedy52 commented:
I believe one alternative is to code that kind of logic in go:
https://github.com/gravitational/teleport/blob/37e52dc8a8e2c60c710cea3d3c8e78dedbb18d5d/lib/auth/auth_with_roles.go#L5842-L5850

So on user create/update/delete operations, if it's built-in okta service, then the user must have that label. Probably want to make sure users created/updated in other ways do NOT have that label.

tcsc commented:
Ah, so we _can_ do it - it just needs custom logic at the access check, rather than being part of our normal RBAC system. I'll have a play and see if I can do similar.

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/add-okta-user-rw

tcsc committed:
Restricts User RBAC operations on the built-in Okta role

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

greedy52 commented on lib/auth/auth_with_roles.go:
```diff
@@ -3446,6 +3490,16 @@ func (a *ServerWithRoles) CompareAndSwapUser(ctx context.Context, new, existing
+			return trace.BadParameter("Okta service must supply okta origin")
+		}
+
+		if err := a.isOktaUser(ctx, new.GetName()); err != nil {
+			return trace.Wrap(err)
+		}
```
```suggestion
		if !hasOriginOkta(existing) {
			return trace.BadParameter("Okta service must supply okta origin")
		}
```

greedy52 reviewed (commented):
Overall approach LGTM. Like mentioned in other thread, the same logic should be applied to the new user service.

jentfoo commented on lib/auth/auth_with_roles.go:
```diff
@@ -3400,17 +3406,45 @@ func (a *ServerWithRoles) CreateUser(ctx context.Context, user types.User) (type
+
+	if a.hasBuiltinRole(types.RoleOkta) && !hasOriginOkta(user) {
+		return nil, trace.BadParameter("Okta service must supply okta origin")
+	}
```
```suggestion
	} else if ! a.hasBuiltinRole(types.RoleOkta) && hasOriginOkta(user) {
		return nil, trace.BadParameter("Must be Okta service to set Okta origin")
	}
```

This check can be optimized into a single condition if we are ok with changing the error response (`a.hasBuiltinRole(types.RoleOkta) != hasOriginOkta(user)`), but I think we should also be verifying against this condition.

jentfoo commented on lib/auth/auth_with_roles.go:
```diff
@@ -3400,17 +3406,45 @@ func (a *ServerWithRoles) CreateUser(ctx context.Context, user types.User) (type
 
+	if a.hasBuiltinRole(types.RoleOkta) {
+		if !hasOriginOkta(user) {
+			return nil, trace.BadParameter("Okta service must supply okta origin")
```
I feel like this error message is confusing

jentfoo replied:
```suggestion
			return nil, trace.BadParameter("Okta service can only update users of okta origin")
```

jentfoo commented on lib/auth/auth_with_roles.go:
```diff
@@ -3400,17 +3406,45 @@ func (a *ServerWithRoles) CreateUser(ctx context.Context, user types.User) (type
 	}
 
+	if a.hasBuiltinRole(types.RoleOkta) {
+		if !hasOriginOkta(user) {
```
Similar to the other conditional recommendation, we need to verify that someone is not setting an account as Okta without having the Okta role.

tcsc replied:
It's perfectly legitimate for a non-Okta user to Update a user with the Okta origin label. Editing via the UI, for example. We can possibly check that the user is not _adding_ origin label - but that might be getting a bit convoluted.

jentfoo replied:
I am not sure I understand the use case of manually modifying Okta users.

Would that not result in whatever manual changes were done to be overdone if a change occurred on Okta?  Or is there a sync that occurs into Okta?

> We can possibly check that the user is not adding origin label

I think we should at a minium

tcsc replied:
There is a sync process; basically all traits with an `okta/` prefix will be updated to match the upstream Okta profile. An admin should be able to manually add roles to the user and have them preserved.

jentfoo replied:
What triggers this sync?  How do we know the sync will happen in sequence to other updates?

tcsc replied:
Sync runs every _n_ minutes, where _n_ is currently `10`, but we may change it depending on how it interacts with Okta API rate limits.

jentfoo replied:
Since Okta is often driven by automation, I feel like there is a serious risk that a change made on Teleport could easily be undone.  If this is a necessary use case it may be better to understand what changed within the user and commit the change so that we can apply the diffs across updates.  As an MVP it would be simpler to avoid allowing Okta users to be managed outside of Okta.

jentfoo commented on lib/auth/auth_with_roles.go:
```diff
@@ -3420,6 +3454,16 @@ func (a *ServerWithRoles) UpsertUser(ctx context.Context, u types.User) (types.U
 		return nil, trace.Wrap(err)
 	}
 
+	if a.hasBuiltinRole(types.RoleOkta) {
```
Considering this conditional is repeated it might be worth putting into a function

jentfoo reviewed (commented):
Overall looks workable, just a few suggestions for improvement

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

tcsc committed:
Address review comments

 - Consolidates update checking into a single function
 - Adds implementation for new user service

tcsc committed:
remove unused code

greedy52 commented on lib/auth/users/usersv1/okta_test.go:
```diff
@@ -0,0 +1,299 @@
+	user, err := types.NewUser(t.Name())
+	require.NoError(t, err)
+
+	m := user.GetMetadata()
+	m.Labels = map[string]string{types.OriginLabel: types.OriginOkta}
+	user.SetMetadata(m)
```
nit:
```suggestion
	user.SetOrigin(types.OriginOkta)
```

greedy52 commented on lib/auth/auth_with_roles.go:
```diff
@@ -3410,10 +3416,38 @@ func (a *ServerWithRoles) CreateUser(ctx context.Context, user types.User) (type
+		}
+	} else {
+		if hasOriginOkta(user) {
+			return nil, trace.BadParameter("Must be Okta service to set Okta origin")
```
nit: these two `BadParameter` should be consistent
```suggestion
			return nil, trace.BadParameter("Must be Okta service to set okta origin")
```
may quote "okta"

greedy52 commented on lib/auth/users/usersv1/okta.go:
```diff
@@ -0,0 +1,77 @@
+// existing user resource and the new value it will take after any update
+// completes. If either of these records is missing the label, this function
+// returns an error.
+func (s *Service) checkOktaUpdateAccess(ctx context.Context, authzCtx *authz.Context, new types.User, existingUsername string) error {
```
nit: I don't think we need `existingUsername` here, can just get from `new` (types.User)

greedy52 commented on lib/auth/users/usersv1/okta.go:
```diff
@@ -0,0 +1,77 @@
+func isOktaWriteableUserResource(ctx context.Context, cache Cache, username string) error {
+	targetUser, err := cache.GetUser(ctx, username, false)
+	switch {
+	case trace.IsNotFound(err):
+		// If no such user exists, then we will treat it as writable. Otherwise
+		// we won't be able to create the user on an upsert.
+		return nil
```
nit: I think this logic is OK but personally I would prefer to return the NotFound and let the caller decide what to do.

tcsc commented on lib/auth/auth_with_roles.go:
```diff
@@ -7198,3 +7260,10 @@ func verbsToReplaceResourceWithOrigin(stored types.ResourceWithOrigin) []string
 	return verbs
 }
+
+func hasOriginOkta(r types.Resource) bool {
+	if resourceOrigin, present := r.GetMetadata().Labels[types.OriginLabel]; present {
+		return resourceOrigin == types.OriginOkta
+	}
+	return false
+}
```
Did not know that. Will definitely simplify.

tcsc commented on lib/auth/auth_with_roles.go:
```diff
@@ -3461,6 +3519,10 @@ func (a *ServerWithRoles) CompareAndSwapUser(ctx context.Context, new, existing
 		return trace.Wrap(err)
 	}
 
+	if err := a.checkOktaUpdateAccess(ctx, new, existing.GetName()); err != nil {
+		return trace.Wrap(err)
+	}
```
Given that the intent for that check was to prevent the Okta server annexing a non-Okta user, I think that works:

| `existing` origin | Backend record origin | Access   | Consistent | Notes |
|-------------------|-----------------------|----------|------------|-------|
| Okta              | Okta                  | Approved | Yes        |       |
| Okta              | `$SOMETHING_ELSE`     | Denied   | Yes        | Update will be rejected by `CompareAndSwap` because `existing` and backend values are different      |
| `$SOMETHING_ELSE` | Okta                  | Denied   | Yes        | Update rejected by okta origin check |
| `$SOMETHING_ELSE` | `$SOMETHING_ELSE`     | Denied   | Yes        | Update rejected by okta origin check |

rosstimothy reviewed (commented):
I don't think that we need to implement this twice, once in auth_with_roles and once in the users service for all RPCs. The auth grpc server calls into the users service for all user CRUD operations except for `Update`.

jentfoo commented on lib/auth/auth_with_roles.go:
```diff
@@ -3453,6 +3495,22 @@ func (a *ServerWithRoles) UpdateAndSwapUser(ctx context.Context, user string, wi
 }
 
+func (a *ServerWithRoles) checkOktaUpdateAccess(ctx context.Context, new types.User, existing string) error {
+	if a.hasBuiltinRole(types.RoleOkta) {
```
I considered the conversation here further: https://github.com/gravitational/teleport/pull/33630#discussion_r1382028202

I still think we need to address the concerns here.  Specifically:
1) As discussed above I think there is concerned about the user access state getting out of sync which could result in Okta clobbering other changes
2) We must validate that the Okta origin is not being added in update.  Otherwise this is a potential bypass to allow non-Okta users to now be modified by the Okta user (with collaboration for the initial update)

tcsc committed:
Refactor Okta access and tests

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/add-okta-user-rw

tcsc committed:
Fix merge detritus

tcsc committed:
Doc updates

tcsc commented:
As per a Slack discussion between @r0mant, @jentfoo & myself, the rules for Okta-origin users are now:

> 1. Only the Teleport Okta service may create an Okta-origin user.
> 2. Okta-origin users will only ever have the `requester` role.
> 3. Only the Teleport Okta service may modify an Okta-origin user. If a Teleport admin wants to grant an Okta-origin user long lived access to something, they will use an Access List rather than a direct role assignment.
> 4. Anyone with User RW can delete an Okta-origin user, otherwise there is no recourse for a Teleport admin to clean up obsolete users if an Okta integration is deleted. If a "live" (i.e. someone with a valid upstream Okta account) Okta-origin user is deleted, it will simply be re-created on the next Okta sync.

I will update the code and test cases to match and ask for a final round of review.

tcsc committed:
Hopefully last pass

tcsc committed:
Linter appeasement

tcsc committed:
Merge branch 'master' into tcsc/add-okta-user-rw

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -7213,3 +7251,77 @@ func verbsToReplaceResourceWithOrigin(stored types.ResourceWithOrigin) []string
+
+	switch {
+	case isOktaService && !hasOktaOrigin:
+		return trace.BadParameter(`Okta service must supply "okta" origin`)
```
Nit: Can't we just force-set the origin label if it's an Okta service that's creating the user?

tcsc replied:
Definitely possible, but I'm always a bit leery of silently correcting a caller mistake. That said, it does feel a bit like hoop jumping to force it.

r0mant commented on lib/auth/users/usersv1/okta.go:
```diff
@@ -0,0 +1,89 @@
+// without the Okta role may supply any origin value *other than* okta
+// (including nil).
+// Returns an error if the user origin value is "inappropriate".
+func checkOktaOrigin(authzCtx *authz.Context, user types.User) error {
```
Looks like functions in this file are duplicated from auth_with_roles.go. Can we reuse?

tcsc replied:
The user management in `auth_with_roles.go` is slated to be deleted in 16.0, and this will go with it, so I didn't spend much time trying to make something that worked for both cases. Happy to try and merge if we think its worthwhile.

tcsc replied:
Actually, done

tcsc committed:
Merge remote-tracking branch 'origin/master' into tcsc/add-okta-user-rw

tcsc committed:
Consilidated checks

jentfoo reviewed (approved):
Thank you for the additional work here, the latest changes look good to me!

tcsc committed:
Update auth_with_roles_okta_rbac_test.go

docs
